resources/ManageAPICredentials.php

```diff
@@ -0,0 +1,239 @@
+            if ($options->getDelete()) {
+                deleteCreds($entityManager, $creds);
+            }
+            if ($options->getWarn()) {
+                warnUsers($creds, $threshold);
+            }
+        } else {
+            reportDryRun($creds, $options->getWarn() ? "sending warning email" : "deletion");
+        }
+    } else {
+        if ($options->getDryRun()) {
+            print("Dry run: No matching credentials found.\n");
+        }
+    }
+} catch (InvalidArgumentException $except) {
+    $options->usage($except->getMessage());
+}
+
+function reportDryRun(array $creds, $text)
+{
+    print("Dry run: Found " . count($creds) . " credentials.\n");
+    foreach ($creds as $api) {
+        dumpCredential($api, $text);
+    }
+}
+function dumpCredential(APIAuthentication $api, $text)
+{
+    print("Dry run: Processing credential id " . $api->getId() . " for $text\n"
+        . "         Identifier: " . $api->getIdentifier() . "\n"
+        . "         User email: " . $api->getUser()->getEmail() . "\n"
+        . "         Site:       " . $api->getParentSite()->getShortName() . "\n"
+        . "         Last used:  " . $api->getLastUseTime()
+                                ->format("Y-m-d\\TH:i:sO")  // DateTimeInterface::ISO8601
+        . "\n"
+    );
+}
+function deleteCreds($entityManager, $creds)
+{
+    $serv = new APIAuthenticationService();
+    $serv->setEntityManager($entityManager);
+
+    /* @var $api APIAuthentication */
+    foreach ($creds as $api) {
+        $serv->deleteAPIAuthentication($api);
+    }
+}
+function warnUsers($creds, $deletionThreshold)
+{
+    /* @var $api APIAuthentication */
+    foreach ($creds as $api) {
+        $lastUsed = $api->getLastUseTime();
+
+        $timeDiff = (new DateTime())->diff($lastUsed);
+        $lastUseMonths = $timeDiff->format('%m');
+
+        sendWarningEmail($api, intval($lastUseMonths), $deletionThreshold);
+    }
+}
+
+function getCreds($entityManager, $threshold)
+{
+    $qbl = $entityManager->createQueryBuilder();
+
+    $qbl->select('cred')
+        ->from('APIAuthentication', 'cred')
+        ->where('cred.lastUseTime < :threshold');
+
+    $timeThresh = new DateTime();
+    $timeThresh->sub(new DateInterval("P" . $threshold . "M"));
+    $qbl->setParameter('threshold', $timeThresh->format('Y-m-d H:i:s'));
+
+    $creds = $qbl->getQuery()->getResult();
+
+    return $creds;
+}
+function sendWarningEmail(\APIAuthentication $api, $elapsedMonths, $deletionThreshold)
+{
+    $user = $api->getUser();
+    $userEmail = $user->getEmail();
+    $siteName = $api->getParentSite()->getShortName();
+    $siteEmail = $api->getParentSite()->getEmail();
+
+    $headers = "From: GOCDB <gocdb-admins@mailman.egi.eu>";
+    $headers .= "\r\nCC: $siteName <$siteEmail>";
+    $subject = "GOCDB: Site API credential deletion notice";
+
+    $body = "Dear " . $user->getForename() . ",\n\n" .
+        "The API credential associated with the following identifiers " .
```

Comment:
API credentials associated to accounts, not identifers, right?

Maybe here it's better to to say ""The API credential added by user X registered at site ..."

Reply:
and then remove `$body .= "Identifier: " . $api->getIdentifier() . "\n";`

Reply:
ah sorry, I'm being dense - it's the API credential identifier you're displaying 🤦 

Maybe it would be good to still say which user added it? So if you are actioning the email because of the cc you have a bit more context?

Reply:
OK will add user email for context.